pkg/util/bulk/aggregator_stats.go

```diff
@@ -0,0 +1,154 @@
+			component.SQLInstanceID.String(), component.FlowID.String()))
+		for name, event := range nameToEvent {
+			// Write a proto file per tag. This machine-readable file can be consumed
+			// by other places we want to display this information egs: annotated
+			// DistSQL diagrams, DBConsole etc.
+			filename := fmt.Sprintf("%s/%s",
+				component.SQLInstanceID.String(), asOf)
+			msg, err := protoreflect.DecodeMessage(name, event)
+			if err != nil {
+				clusterWideSummary.WriteString(fmt.Sprintf("invalid protocol message: %v", err))
+				// If we failed to decode the event write the error to the file and
+				// carry on.
+				continue
+			}
+
+			if err := jobs.WriteProtobinExecutionDetailFile(ctx, filename, msg, db, jobID); err != nil {
+				return err
+			}
+
+			// Construct a single text file that contains information on a per-node
+			// basis as well as a cluster-wide aggregate.
+			clusterWideSummary.WriteString(fmt.Sprintf("# %s\n", name))
+
+			aggEvent := msg.(TracingAggregatorEvent)
+			clusterWideSummary.WriteString(aggEvent.String())
+			clusterWideSummary.WriteString("\n")
+
+			if _, ok := clusterWideAggregatorStats[name]; ok {
+				clusterWideAggregatorStats[name].Combine(aggEvent)
+			} else {
+				clusterWideAggregatorStats[name] = aggEvent
+			}
+		}
+	}
+
+	for tag, event := range clusterWideAggregatorStats {
+		clusterWideSummary.WriteString("## Cluster-wide\n\n")
+		clusterWideSummary.WriteString(fmt.Sprintf("# %s\n", tag))
+		clusterWideSummary.WriteString(event.String())
+	}
+
+	filename := fmt.Sprintf("aggregatorstats.%s.txt", asOf)
+	return jobs.WriteExecutionDetailFile(ctx, filename, clusterWideSummary.Bytes(), db, jobID)
+}
+
+// AggregateTracingStats listens for AggregatorEvents on a channel and
+// periodically processes them to human and machine-readable file formats that
+// are persisted in the system.job_info table. These files can then be consumed
+// for improved observability into the job's execution.
+//
+// This method does not return until the passed in channel is closed or an error
+// is encountered.
+func AggregateTracingStats(
+	ctx context.Context,
+	jobID jobspb.JobID,
+	st *cluster.Settings,
+	db isql.DB,
+	tracingAgg chan *execinfrapb.TracingAggregatorEvents,
+) error {
+	perNodeAggregatorStats := make(map[execinfrapb.ComponentID]map[string][]byte)
+
+	// AggregatorEvents are periodically received from each node in the DistSQL
+	// flow.
+	flushTimer := timeutil.NewTimer()
+	defer flushTimer.Stop()
+	flushTimer.Reset(flushTracingAggregatorFrequency.Get(&st.SV))
+
+	var flushOnClose bool
+	for agg := range tracingAgg {
+		flushOnClose = true
+		componentID := execinfrapb.ComponentID{
+			FlowID:        agg.FlowID,
+			SQLInstanceID: agg.SQLInstanceID,
+		}
+
+		// Update the running aggregate of the component with the latest received
+		// aggregate.
+		perNodeAggregatorStats[componentID] = agg.Events
+
+		select {
+		case <-flushTimer.C:
```

Comment:
I wonder if something like `if time.Since(lastFlushTime) > flushTracingAggregatorFrequency.Get(&st.SV)` would be clearer. But this has grown on me on a second reading.